tools/skbstat.py

```diff
@@ -0,0 +1,996 @@
+        try_probe_list(b, probe_list_output_tcp)
+
+    if proto == 0 or proto == 17:
+        try_probe_list(b, probe_list_output_udp)
+
+def try_probe_drop(b):
+    try_probe(b, "kfree_skb", "trace_skb_drop")
+
+def try_probe_consume(b):
+    try_probe(b, "consume_skb", "trace_skb_consume")
+
+# install probes
+try_probe_drop(b)
+
+if args.input:
+    try_probe_dev_input(b)
+    try_probe_network_input(b, ip_version)
+    try_probe_transport_input(b, ip_version, proto)
+
+if args.output:
+    try_probe_dev_output(b)
+    try_probe_network_output(b, ip_version)
+    try_probe_transport_output(b, ip_version, proto)
+
+if args.consume:
+    try_probe_consume(b)
+
+# get maps
+input_map = b.get_table("input")
+output_map = b.get_table("output")
+
+stack_traces = b.get_table("stack_traces")
+drop_stacks = b.get_table("drop_stacks")
+consume_stacks = b.get_table("consume_stacks")
+
+def sorter(item, probe_list):
+    name = item[0]
+
+    for p in probe_list:
+        if name in p:
+            return p[name][1]
+
+    return 0xffffffff
+
+def input_sorter(item):
+    return sorter(item, probe_list_input)
+
+def output_sorter(item):
+    return sorter(item, probe_list_output)
+
+def print_routines(b, maps, sorter):
+    funcs = []
+    for addr, count in maps.items():
+        funcs.append([b.ksym(addr), sum(count), ""])
+
+    for item in sorted(funcs, key=sorter):
+        print("%-8s %-6d %s" %(" ", item[1], item[0]))
+
+def print_routines_header(direction):
+    print("\n%-8s %-6s %s" % ("TIME", "NUM", "FUNC/STACK"))
+    print("%-8s %-6s [%s]" % (strftime("%H:%M:%S"), " ", direction))
+
+def print_input_routines():
+    print_routines_header("INGRESS")
+    print_routines(b, input_map, input_sorter)
+    input_map.clear()
+
+def print_output_routines():
+    print_routines_header("EGRESS")
+    print_routines(b, output_map, output_sorter)
+    output_map.clear()
+
+def print_backtraces(b, stacks, stack_traces):
+    for stack_id, count in sorted(stacks.items(), 
+            key=lambda kv: sum(kv[1])):
+        print("%-8s %-6d" %(" ", sum(count)))
+        for addr in stack_traces.walk(stack_id.value):
+            sym = b.ksym(addr, show_offset=True)
+            print("%-8s %-6s %s" % (" ", " ", sym))
+
+def print_drop_backtraces():
+    print_routines_header("DROP")
+    print_backtraces(b, drop_stacks, stack_traces)
+
+def print_consume_backtraces():
+    print_routines_header("CONSUME")
+    print_backtraces(b, consume_stacks, stack_traces)
+
+def print_all_backtraces(consume, drop):
+    if drop:
+        print_drop_backtraces()
+    if consume:
+        print_consume_backtraces()
+
+    drop_stacks.clear()
+    consume_stacks.clear()
+    stack_traces.clear()
+
+# read events
+while 1:
+    try:
+        sleep(interval)
+    except KeyboardInterrupt:
+        exit()
+
+    if args.input:
+        print_input_routines()
+
+    if args.output:
+        print_output_routines()
+
+    print_all_backtraces(args.consume, True)
```

Comment:
drop is always true here.

Reply:
Yeah. DROP is the default action here by design. In most cases, people may really care about: "why kernel drops my packet?". So `skbstat` dump drops backtrace by default.

Reply:
I think I can add `--drop` argument, for better output control.